### PR TITLE
Add Zookeeper pagination support for /brokers/topics znode

### PR DESCRIPTION
**Pull request by groelofs**
LIKAFKA-49497: Switch from Apache to LI Zookeeper dependency and add GetAllChildrenPaginated option for /brokers/topics znode (supports 'list topics' responses greater than 1 MB). Feature is controlled by new li.zookeeper.pagination.enable config (default = false).

This includes a new test case and also has been manually tested in the negative sense, but the latter loops infinitely due to a preexisting bug that I'll address in a later PR.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala [hunk 1]
```diff
@@ -198,6 +207,23 @@ class ZooKeeperClient(connectString: String,
...
+            try {
+              topicsList = zooKeeper.getAllChildrenPaginated(path, shouldWatch(request))
+            } catch {
+              case e: KeeperException =>
+                resultCode = e.code
+            }
+            GetChildrenPaginatedResponse(resultCode, path, Option(ctx), Option(topicsList).map(_.asScala).getOrElse(Seq.empty),
+              responseMetadata(sendTimeMs))
+          }
+        }, zkPaginationExecutor).thenAccept(new Consumer[GetChildrenPaginatedResponse] {
```
Comment (hshi2022) [inline core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala:224]: Why do we want to use additional threads zkPaginationExecutor for this request while others do not? Is it because getAllChildrenPaginated only has sync version?

File: core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala [hunk 2]
```diff
@@ -198,6 +207,23 @@ class ZooKeeperClient(connectString: String,
...
         }, ctx.orNull)
+      case GetChildrenPaginatedRequest(path, _, ctx) =>
+        CompletableFuture.supplyAsync(new Supplier[GetChildrenPaginatedResponse]() {
+          def get(): GetChildrenPaginatedResponse = {
+            var topicsList: JList[String] = null 
+            var resultCode: Code = Code.OK
+            try {
+              topicsList = zooKeeper.getAllChildrenPaginated(path, shouldWatch(request))
+            } catch {
+              case e: KeeperException =>
```
Comment (hshi2022) [inline core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala:218]: How about other Exception type? Do we want to handle?

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 3]
```diff
@@ -1158,7 +1161,7 @@ object KafkaConfig {
     "zero. When closing/shutting down an observer, most time is spent on flushing the observed stats. The reasonable timeout should be close to " +
     "the time it takes to flush the stats."
 
-  @nowarn("cat=deprecation")
+  //@nowarn("cat=deprecation")
```
Comment (hshi2022) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: Why comment out this and below annotations? If we want to disable it, we could delete them.

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 4]
```diff
@@ -332,6 +332,7 @@ object Defaults {
   val LiLogCleanerFineGrainedLockEnabled = true
   val LiDropCorruptedFilesEnabled = false
   val LiConsumerFetchSampleRatio = 0.01
+  val LiZookeeperPaginationEnable = false
```
Comment (hshi2022) [inline core/src/main/scala/kafka/server/KafkaConfig.scala:334]: Are we planning to enable this only in clusters that have topic-name-length larger than some threshold? Or do we want to enable it by default? Any performance degradation for calling paginated one?

File: core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala [hunk 5]
```diff
@@ -75,6 +79,11 @@ class ZooKeeperClient(connectString: String,
   private[zookeeper] val reinitializeScheduler = new KafkaScheduler(threads = 1, s"zk-client-${threadPrefix}reinit-")
   private var isFirstConnectionEstablished = false
 
+  // There should never be more than one controller-failover call at a time, but no
+  // guarantees for end-user listTopics() calls, and zookeeper.max.in.flight.requests
+  // appears to be either 10 (default) or infinite, so go with 10 as safe bet:
```
Comment (hshi2022) [inline core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala]: "infinite"? zookeeper.max.in.flight.requests is an int, so between 0 and int.max. What is the possibility that there would be 10 listTopics at the same time? If it is low possibility, feel using a fixed of 10 threads is kind of wasting.

File: core/src/test/scala/unit/kafka/zk/KafkaZkClientTest.scala [hunk 6]
```diff
@@ -1057,6 +1057,51 @@ class KafkaZkClientTest extends ZooKeeperTestHarness {
     assertEquals(Set(topicPartition10, topicPartition11), zkClient.getAllPartitions)
   }
 
+  /**
+   * Create a large number of topic names, such that their total (aggregate) length is more than 1 MB,
+   * and verify that our ZK client can still read them all despite being configured for 1 MB max response
+   * size.
+   *
+   * NOTE THAT THIS LOOPS INDEFINITELY if paginateTopics = false, so a negative test is currently not
```
Comment (hshi2022) [inline core/src/test/scala/unit/kafka/zk/KafkaZkClientTest.scala:1065]: If paginateTopic=false, GetChildrenRequest would be called. Does this mean there is a bug when calling LiZookeeper GetChildrenRequest?

File: core/src/test/scala/unit/kafka/zk/KafkaZkClientTest.scala [hunk 7]
```diff
@@ -1057,6 +1057,51 @@ class KafkaZkClientTest extends ZooKeeperTestHarness {
...
+
+    val numTopics = 8000
+    var sumOfInputTopicNameLengths = 0
+    (0 until numTopics).foreach { j =>
+      if (((j+1) % 1000) == 0) info(s"creating very long topic-name #${j+1}")
+      val name = s"topic-name-of-super-excessive-length-so-as-to-minimize-total-number-of-topics-to-be-created--just-how-lengthy-is-it-I-wonder--here-comes-the-indexing-value-woohoo--${j}"
+      sumOfInputTopicNameLengths += name.length + 4  // 4 = measured ZK overhead per znode
+      zkClient.createRecursive(TopicZNode.path(name))
+    }
+    debug(s"total /brokers/topics getChildren size should be ${sumOfInputTopicNameLengths}")
```
Comment (hshi2022) [inline core/src/test/scala/unit/kafka/zk/KafkaZkClientTest.scala:1096]: Can you put a comment indicating what is sumOfInputTopicNameLengths with the current implementation?

File: core/src/test/scala/unit/kafka/zk/KafkaZkClientTest.scala [hunk 8]
```diff
@@ -1057,6 +1057,51 @@ class KafkaZkClientTest extends ZooKeeperTestHarness {
...
+
+    val topicsList = zkClient.getAllTopicsInCluster()
+    debug(s"got back list of ${topicsList.size} topics")
+    assertEquals(numTopics, topicsList.size)
+
+    assertTrue(sumOfInputTopicNameLengths > (1024*1024))
+    // can't just do topicsList.map here since map of a Set is also a Set, and the _lengths_ of our
+    // topic names are NOT globally unique:
+    val sumOfOutputTopicNameLengths = topicsList.toSeq.map(_.length + 4).sum
```
Comment (hshi2022) [inline core/src/test/scala/unit/kafka/zk/KafkaZkClientTest.scala:1107]: It would also be nice check a couple values in topicsList that are as expected in addition to verify the size of topicsList.

File: core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala (hunk 1 above)
Comment (groelofs) [reply core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala:224]: Exactly, yes.

File: core/src/main/scala/kafka/server/KafkaConfig.scala (hunk 3 above)
Comment (groelofs) [reply core/src/main/scala/kafka/server/KafkaConfig.scala]: I thought it was weird that they're all there but the compiler warns about them being pointless. I was hoping to get some context on why they were there in the first place.

But I'll go ahead and delete them altogether in the next update.

File: core/src/main/scala/kafka/server/KafkaConfig.scala (hunk 4 above)
Comment (groelofs) [reply core/src/main/scala/kafka/server/KafkaConfig.scala:334]: I discussed this a bit in one of our team meetings: there's potentially a performance cost, though likely quite small. There's also just the basic risk of using a new API that hasn't seen much real-world testing. So yes, I was thinking to enable it only for samza-internal1 initially, and then perhaps others if/when they approach the 800 KB mark.

File: core/src/test/scala/unit/kafka/zk/KafkaZkClientTest.scala (hunk 6 above)
Comment (groelofs) [reply core/src/test/scala/unit/kafka/zk/KafkaZkClientTest.scala:1065]: Correct. More precisely, the bug is in our ZooKeeperClient class's handling of the error-response. Arguably there's also a bug in Zookeeper itself in that it apparently uses a SessionExpiredException (retryable) instead of an explicitly non-retryable exception (jute.maxbuffer isn't going to fix itself no matter how many times you retry.)

I'll fix the bug on our side in a followup (via a counter and configurable max).

File: core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala (hunk 5 above)
Comment (groelofs) [reply core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala]: Yeah, I'm not actually sure why I wrote that...the live configured value is 10 in my dev deployment, and 0 is disallowed by the code, so it doesn't count as "infinite."

Anyway, the probability is definitely low, but one could imagine a service doing a 100% bounce or deploy, with every instance coming up and trying to do one of these calls. What would you suggest as an alternative? While it's true there's some overhead to extra threads, if they're not doing anything, it's essentially zero CPU, and memory is mostly the stack, I think (which may default to 10 MB, but that's tiny relative to our deployed setup).

File: core/src/main/scala/kafka/zk/KafkaZkClient.scala [hunk 9]
```diff
@@ -51,8 +51,8 @@ import scala.collection.{Map, Seq, mutable}
  * easier to migrate away from `ZkUtils` (since removed). We should revisit this. We should also consider whether a
  * monolithic [[kafka.zk.ZkData]] is the way to go.
  */
-class KafkaZkClient private[zk] (zooKeeperClient: ZooKeeperClient, isSecure: Boolean, time: Time) extends AutoCloseable with
-  Logging with KafkaMetricsGroup {
+class KafkaZkClient private[zk] (zooKeeperClient: ZooKeeperClient, isSecure: Boolean, time: Time, paginateTopics: Boolean)
```
Comment (lmr3796) [inline core/src/main/scala/kafka/zk/KafkaZkClient.scala]: minor optional nit:

Can we follow the line split pattern as the other classes, placing each parameter on 1 line?

`ZkAdminManager` does it this way
```scala
class ZkAdminManager(val config: KafkaConfig,
                     val metrics: Metrics,
                     val metadataCache: MetadataCache,
                     val zkClient: KafkaZkClient,
                     val controller: KafkaController) extends Logging with KafkaMetricsGroup {
```

File: core/src/main/scala/kafka/zk/KafkaZkClient.scala [hunk 10]
```diff
@@ -546,7 +546,8 @@ class KafkaZkClient private[zk] (zooKeeperClient: ZooKeeperClient, isSecure: Boo
    */
   def getAllTopicsInCluster(registerWatch: Boolean = false): Set[String] = {
     val getChildrenResponse = retryRequestUntilConnected(
-      GetChildrenRequest(TopicsZNode.path, registerWatch))
+      if (paginateTopics) GetChildrenPaginatedRequest(TopicsZNode.path, registerWatch) else
```
Comment (lmr3796) [inline core/src/main/scala/kafka/zk/KafkaZkClient.scala]: nit: put else part on a separate line

File: core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala (hunk 5 above)
Comment (lmr3796) [reply core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala]: I agree with 10 threads are not a significant pressure, especially when compared with the maintenance cost of debugging blocked requests on this, I'd say configured to the imaginable quota.

File: core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala (hunk 2 above)
Comment (lmr3796) [reply core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala:218]: I traced the code. 

For other un-checked exceptions, it would end up `CompletionException`, and the upper level catches those
https://github.com/linkedin/kafka/blob/21cece149e5668dfc4f40ee61fa1741d71d51270/core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala#L163

File: core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala [hunk 11]
```diff
@@ -569,6 +601,11 @@ sealed abstract class AsyncResponse {
   def metadata: ResponseMetadata
 }
 
+sealed trait GetChildrenResponseNoStat extends AsyncResponse {
```
Comment (lmr3796) [inline core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala:604]: Why do we want to customize to a NoStat?

File: core/src/test/scala/unit/kafka/zk/EmbeddedZookeeper.scala [hunk 12]
```diff
@@ -40,7 +48,8 @@ class EmbeddedZookeeper() extends Logging {
   val logDir = TestUtils.tempDir()
   val tickTime = 800 // allow a maxSessionTimeout of 20 * 800ms = 16 secs
 
-  System.setProperty("zookeeper.forceSync", "no")  //disable fsync to ZK txn log in tests to avoid timeout
+  System.setProperty("zookeeper.forceSync", "no")  // disable fsync to ZK txn log in tests to avoid timeout
```
Comment (lmr3796) [inline core/src/test/scala/unit/kafka/zk/EmbeddedZookeeper.scala]: optional nit: I personally prefer to leave this kind of styling changes as in the upstream.  This makes future rebase easier.

File: core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala (hunk 5 above)
Comment (hshi2022) [reply core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala]: Yes, I agree that 10 threads are not expensive. However, if we create unnecessary threads here and there, the accumulated usage could be huge. I am wondering whether controller has a shared threadpool that we could take threads from. If not, if the request rate is low, we could get a new thread each time for a new paginated request, so we do not need to have 10 threads waiting for most of the time. If the request rate is high and performance requirement is high, I guess having 10 threads should be fine.

File: core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala (hunk 2 above)
Comment (groelofs) [reply core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala:218]: No, the caller handles all Throwables for all of the APIs. All we need here is the error code from any ZK server errors:

`@throws KeeperException if the server signals an error with a non-zero error code`

File: core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala (hunk 11 above)
Comment (groelofs) [reply core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala:604]: The paginated call doesn't support the Stat argument, and there are a handful of getChildren() callers that actually do make use of it. (My first thought was to get rid of the argument altogether, but no joy.) Having a shared superclass (or whatever it's called in Scala) saves a little bit of code in getAllTopicsInCluster() in KafkaZkClient.

File: core/src/test/scala/unit/kafka/zk/EmbeddedZookeeper.scala (hunk 12 above)
Comment (groelofs) [reply core/src/test/scala/unit/kafka/zk/EmbeddedZookeeper.scala]: Oops, good point. I usually manage to avoid giving in to my OCD impulses. :-D

File: core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala (hunk 5 above)
Comment (lmr3796) [reply core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala]: Controller has a thread pool dedicated for requests.  However this ZK part is a hidden component encapsulated underneath.  

If we'd like to control it better, maybe pull that to a upper level config instead of hard coding.

File: core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala (hunk 11 above)
Comment (lmr3796) [reply core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala:604]: Makes sense.

**Review by lmr3796 — APPROVED** [review]
LGTM
